Relationship with UUID + DTO

<!--
- Please follow the issue template below for bug reports.
- If you have a support request rather than a bug, please use [Stack Overflow](http://stackoverflow.com/questions/tagged/jhipster) with the JHipster tag.
- For bug reports it is mandatory to run the command `jhipster info` in your project's root folder, and paste the result here.
- Tickets opened without any of these pieces of information will be **closed** without any explanation.
-->

##### **Overview of the issue**

When generating a relationship with UUID + DTO (mapstruct), the generated project didn't compile.

##### **Motivation for or Use Case**

Here the stacktrace:

```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/pgrimaud/tmp/04-bug-uuid-dto/src/main/java/io/github/pascalgrimaud/toto/service/mapper/AlphaMapper.java:[16,14] Can't map property "java.util.UUID alpha.beta.uuidCode" to "java.lang.String betaUuidCode". Consider to declare/implement a mapping method: "java.lang.String map(java.util.UUID value)".
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  5.666 s
[INFO] Finished at: 2020-01-31T10:17:49+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project jhipster: Compilation failure
[ERROR] /home/pgrimaud/tmp/04-bug-uuid-dto/src/main/java/io/github/pascalgrimaud/toto/service/mapper/AlphaMapper.java:[16,14] Can't map property "java.util.UUID alpha.beta.uuidCode" to "java.lang.String betaUuidCode". Consider to declare/implement a mapping method: "java.lang.String map(java.util.UUID value)".
[ERROR] 

```

##### **Reproduce the error**

- generate a project with all default options
- use this JDL:

```
/**
 * The Alpha entity.
 * @author A true hipster
 */
entity Alpha {
  /** fieldName */
  name String
}

/**
 * The Beta entity.
 * @author A true hipster
 */
entity Beta {
  /** fieldName */
  name String
  uuidCode UUID
}

/**
 * Many to one relationship.
 */
relationship ManyToOne {
  Alpha{beta(uuidCode)} to Beta
}

dto * with mapstruct

```

##### **Related issues**


##### **Suggest a Fix**

Probably something to do in this line:
https://github.com/jhipster/generator-jhipster/blob/master/generators/entity-server/templates/src/main/java/package/service/dto/EntityDTO.java.ejs#L142

But other checks need to be done (tests / integration tests etc)

##### **JHipster Version(s)**

v6.6.0 and current master

##### **JHipster Version(s)**

```
jhipster@0.0.1-SNAPSHOT /home/pgrimaud/tmp/04-bug-uuid-dto
└── (empty)

```


##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**


<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "io.github.pascalgrimaud.toto",
      "nativeLanguage": "en"
    },
    "jhipsterVersion": "6.6.0",
    "applicationType": "monolith",
    "baseName": "jhipster",
    "packageName": "io.github.pascalgrimaud.toto",
    "packageFolder": "io/github/pascalgrimaud/toto",
    "serverPort": "8080",
    "authenticationType": "jwt",
    "cacheProvider": "ehcache",
    "enableHibernateCache": true,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": false,
    "buildTool": "maven",
    "enableSwaggerCodegen": false,
    "jwtSecretKey": "bXktc2VjcmV0LXRva2VuLXRvLWNoYW5nZS1pbi1wcm9kdWN0aW9uLWFuZC10by1rZWVwLWluLWEtc2VjdXJlLXBsYWNl",
    "embeddableLaunchScript": false,
    "useSass": true,
    "clientPackageManager": "npm",
    "clientFramework": "angularX",
    "clientTheme": "none",
    "clientThemeVariant": "",
    "creationTimestamp": 1580459341222,
    "testFrameworks": [],
    "jhiPrefix": "jhi",
    "entitySuffix": "",
    "dtoSuffix": "DTO",
    "otherModules": [],
    "enableTranslation": true,
    "nativeLanguage": "en",
    "languages": ["en"],
    "blueprints": []
  }
}

</pre>
</details>


##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
/**
 * The Alpha entity.\n@author A true hipster
 */
entity Alpha {
  /**
   * fieldName
   */
  name String
}
/**
 * The Beta entity.\n@author A true hipster
 */
entity Beta {
  /**
   * fieldName
   */
  name String,
  uuidCode UUID
}
relationship ManyToOne {
  Alpha{beta(uuidCode)} to Beta
}

dto Alpha, Beta with mapstruct
service Alpha, Beta with serviceClass

</pre>
</details>


##### **Environment and Tools**

openjdk version "11.0.6" 2020-01-14
OpenJDK Runtime Environment (build 11.0.6+10-post-Ubuntu-1ubuntu118.04.1)
OpenJDK 64-Bit Server VM (build 11.0.6+10-post-Ubuntu-1ubuntu118.04.1, mixed mode, sharing)

git version 2.25.0

node: v12.14.0

npm: 6.13.4

yeoman: 3.1.1

yarn: 1.21.1

Docker version 18.03.0-ce, build 0520e24

docker-compose version 1.21.2, build a133471

--- a/generators/entity-server/templates/src/main/java/package/service/mapper/EntityMapper.java.ejs
+++ b/generators/entity-server/templates/src/main/java/package/service/mapper/EntityMapper.java.ejs
@@ -18,7 +18,7 @@
 -%>
 package <%= packageName %>.service.mapper;
 
-<%_ 
+<%_
 let existingMappings = [];
 let uuidMapMethod = false;
 let byteMapMethod = false;
@@ -28,12 +28,17 @@ for (idx in relationships) {
         if (!existingMappings.includes(relationships[idx].otherEntityNameCapitalized) && asEntity(relationships[idx].otherEntityNameCapitalized) !== asEntity(entityClass)) {
             existingMappings.push(relationships[idx].otherEntityNameCapitalized);
         }
-        let otherEntityField = getExistingEntities().find(e => e.name === relationships[idx].otherEntityNameCapitalized).definition.fields.find(f => f.fieldName === relationships[idx].otherEntityField);
-        if (otherEntityField.fieldType === 'UUID') {
-            uuidMapMethod = true;
-        }
-        if (otherEntityField.fieldType === 'byte[]') {
-            byteMapMethod = true;
+        let otherEntity = getExistingEntities().find(e => e.name === relationships[idx].otherEntityNameCapitalized);
+        if (otherEntity) {
+            let otherEntityField = otherEntity.definition.fields.find(f => f.fieldName === relationships[idx].otherEntityField);
+            if (otherEntityField) {
+                if (otherEntityField.fieldType === 'UUID') {
+                    uuidMapMethod = true;
+                }
+                if (otherEntityField.fieldType === 'byte[]') {
+                    byteMapMethod = true;
+                }
+            }
         }
     }
 }

--- a/generators/entity-server/templates/src/main/java/package/service/mapper/EntityMapper.java.ejs
+++ b/generators/entity-server/templates/src/main/java/package/service/mapper/EntityMapper.java.ejs
@@ -18,21 +18,41 @@
 -%>
 package <%= packageName %>.service.mapper;
 
+<%_ 
+let existingMappings = [];
+let uuidMapMethod = false;
+let byteMapMethod = false;
+for (idx in relationships) {
+    if ((relationships[idx].relationshipType === 'many-to-many' && relationships[idx].ownerSide === true) || relationships[idx].relationshipType === 'many-to-one' || (relationships[idx].relationshipType === 'one-to-one' && relationships[idx].ownerSide === true)) {
+        // if the entity is mapped twice, we should implement the mapping once
+        if (!existingMappings.includes(relationships[idx].otherEntityNameCapitalized) && asEntity(relationships[idx].otherEntityNameCapitalized) !== asEntity(entityClass)) {
+            existingMappings.push(relationships[idx].otherEntityNameCapitalized);
+        }
+        let otherEntityField = getExistingEntities().find(e => e.name === relationships[idx].otherEntityNameCapitalized).definition.fields.find(f => f.fieldName === relationships[idx].otherEntityField);
+        if (otherEntityField.fieldType === 'UUID') {
+            uuidMapMethod = true;
+        }
+        if (otherEntityField.fieldType === 'byte[]') {
+            byteMapMethod = true;
+        }
+    }
+}
+_%>
+
 import <%= packageName %>.domain.*;
 import <%= packageName %>.service.dto.<%= asDto(entityClass) %>;
 
 import org.mapstruct.*;
+<%_ if (uuidMapMethod) { _%>
+
+import java.util.Objects;
+import java.util.UUID;
+<%_ } _%>
 
 /**
  * Mapper for the entity {@link <%= asEntity(entityClass) %>} and its DTO {@link <%= asDto(entityClass) %>}.
  */
-@Mapper(componentModel = "spring", uses = {<% var existingMappings = [];
-  for (idx in relationships) {
-    if ((relationships[idx].relationshipType === 'many-to-many' && relationships[idx].ownerSide === true)|| relationships[idx].relationshipType === 'many-to-one' ||(relationships[idx].relationshipType === 'one-to-one' && relationships[idx].ownerSide === true)) {
-      // if the entity is mapped twice, we should implement the mapping once
-      if (!existingMappings.includes(relationships[idx].otherEntityNameCapitalized) && asEntity(relationships[idx].otherEntityNameCapitalized) !== asEntity(entityClass)) {
-          existingMappings.push(relationships[idx].otherEntityNameCapitalized);
-      } } } %><%= existingMappings.map(otherEntityNameCapitalized => otherEntityNameCapitalized + 'Mapper.class').join(', ') %>})
+@Mapper(componentModel = "spring", uses = {<%= existingMappings.map(otherEntityNameCapitalized => otherEntityNameCapitalized + 'Mapper.class').join(', ') %>})
 public interface <%= entityClass %>Mapper extends EntityMapper<<%= asDto(entityClass) %>, <%= asEntity(entityClass) %>> {
 
 <%_
@@ -98,4 +118,16 @@ _%>
         return <%= asEntity(entityInstance) %>;
     }
 <%_ } _%>
+    <%_ if (uuidMapMethod) { _%>
+
+    default String map(UUID value) {
+        return Objects.toString(value, null);
+    }
+    <%_ } _%>
+    <%_ if (byteMapMethod) { _%>
+
+    default String map(byte[] value) {
+        return new String(value);
+    }
+    <%_ } _%>
 }

